src/providers/route.js

```diff
@@ -0,0 +1,244 @@
+
+    _compileUrlPattern(urlPattern, config) {
+      let match;
+      let compiledUrl;
+      urlPattern = this._escapeRegexSpecialCharacters(urlPattern);
+      urlPattern = this._ensureOptionalTrailingSlash(urlPattern);
+
+      let tokenRegex = /\{([A-Za-z\._0-9]+)\}/g;
+      let urlRegex = urlPattern;
+
+      if (!config.partialMatch) {
+        urlRegex = `^${urlRegex}$`;
+      }
+
+      let tokenList = [];
+
+      while ((match = tokenRegex.exec(urlPattern)) !== null) {
+        let token = tokens[match[1]];
+        tokenList.push(token);
+        urlRegex = urlRegex.replace(match[0], `(${types[token.type].regex.source})`);
+      }
+
+      urlRegex.replace('.', '\\.');
+
+      return compiledUrl = {
+        regex: new RegExp(urlRegex, 'i'),
+        tokens: tokenList
+      };
+    },
+
+    _ensureOptionalTrailingSlash(str) {
+      if (str.match(/\/$/)) {
+        return str.replace(/\/$/, '/?');
+      }
+      return str + '/?';
+    },
+
+    _escapeRegexSpecialCharacters(str) {
+      return str.replace(/[\-\[\]\/\(\)\*\+\?\\\^\$\|]/g, "\\$&");
+    },
+
+    $get($location, State, $injector, $q) {
+
+      // When getting a new instance of the service (only done once), we need to iterate over the urlWriters and turn
+      // them into methods that invoke the REAL urlWriter, but providing dependency injection to it, while also
+      // giving it the data that the callee passes in.
+
+      // The reason we have to do this here is because we don't have access to the $injector back in the routeProvider.
+
+      _.forIn(urlWriters, (writer, writerName) =>
+        urlWriters[writerName] = function(data) {
+          if (data == null) { data = {}; }
+          let locals = {UrlData: data};
+          return $injector.invoke(writer, {}, locals);
+        }
+      );
+
+      let flashStates = [];
+
+      let service = {
+        readyDeferred: $q.defer(),
+
+        match(urlToMatch) {
+          for (let url of Array.from(urls)) {
+            var match;
+            if ((match = url.compiledUrl.regex.exec(urlToMatch)) !== null) {
+              return {url, regexMatch: match};
+            }
+          }
+          return null;
+        },
+
+        extractData(match, searchData) {
+          if (searchData == null) { searchData = undefined; }
+          let defaults = this.extractDefaultData(match);
+          let path = this.extractPathData(match);
+          searchData = this.extractSearchData(searchData);
+          return ObjectHelper.default(searchData, path, defaults);
+        },
+
+        extractSearchData(searchData) {
+          if (searchData == null) { searchData = $location.search(); }
+          let data = _.clone(searchData);
+          let newData = {};
+
+          for (let key in data) {
+            var value = data[key];
+            let targetKey = _.findKey(tokens, { searchAlias: key });
+            if (targetKey == null) { targetKey = key; }
+
```

Comment:
What's this guard stuff? Can you tidy up this block so that the code is more readable